## Problem 100: Arranged probability

If a box contains twenty-one coloured discs, composed of fifteen blue discs and six red discs, and two discs were taken at random, it can be seen that the probability of taking two blue discs, $P(BB) = (15/21)×(14/20) = 1/2$.

The next such arrangement, for which there is exactly $50\%$ chance of taking two blue discs at random, is a box containing eighty-five blue discs and thirty-five red discs.

By finding the first arrangement to contain over $10^{12} = 1,000,000,000,000$ discs in total, determine the number of blue discs that the box would contain.

## Answer:

This problem can be translated to find integer $N$, $n$ which satisfy the following equation:

$$
\begin{align}
\frac{n}{N} \times \frac{n-1}{N-1} = \frac{1}{2} \Rightarrow 2 \times n \times (n-1) = N \times (N-1)
\end{align}
$$

Straightforward, I run a loop to find the $N$ and $n$. 

In [1]:
import math

In [2]:
def find_arr(start):
    '''
    Find the arrangement of discs whose P(BB) = 1/2 by trial one-by-one from start

    Args:
        start (int): The number over which to find the next arrangement

    Returns:
        The total number of discs and the number of blue discs
    '''
    N = math.floor(start)
    n = math.floor(0.7 * N)
    t1 = 0
    t2 = 0
    while 2 * n * (n-1) != N * (N - 1):
        if 2 * n * (n-1) < N * (N - 1):
            n = n + 1
            t1 = t1 + 1 
        else:
            N = N + 1
            t2 = t2 + 1
            
    t = t1 + t2 
        
    print ("iterate", t1, "+", t2, "=", t,"times")

    return N, n

In [3]:
print(find_arr(1e2))
print(find_arr(1e4))
print(find_arr(1e6))

iterate 15 + 20 = 35 times
(120, 85)
iterate 9731 + 13661 = 23392 times
(23661, 16731)
iterate 2612555 + 3684660 = 6297215 times
(4684660, 3312555)


It works well when number is small, but after $10^6$ it will be tooooo time-consuming.

Actually, the equation is essentially a quadratic Diophantine equation, and there is an analytical solution to this equation, which turns out to be a pair of recursive expressions.

$$
\begin{align}
N_{i+1} &= 4 \times n_i + 3 \times N_i - 3 \\
n_{i+1} &= 3 \times n_i + 2 \times N_i - 2 \\
\end{align}
$$

So, I can run another loop to find the answer, which is way much faster!

In [4]:
def find_arr2(start):
    '''
    Find the arrangement of discs whose P(BB) = 1/2 by recursive function

    Args:
        start (int): The number over which to find the next arrangement

    Returns:
        The total number of discs and the number of blue discs
    '''
    N = 21
    n = 15
    t = 0
    while N < start:
        N_temp = N
        n_temp = n
        N = 4 * n_temp + 3 * N_temp - 3
        n = 3 * n_temp + 2 * N_temp - 2  
        t = t + 1
        
    print ("iterate", t, "times")
        
    return N, n

In [5]:
print(find_arr2(1e2))
print(find_arr2(1e4))
print(find_arr2(1e6))

iterate 1 times
(120, 85)
iterate 4 times
(23661, 16731)
iterate 7 times
(4684660, 3312555)


For this problem, we should input $10^{12}$. For `find_arr`, it may take hours (or even days!) to find the answer, but `find_arr2` gives the answer just within a second!

In [6]:
print(find_arr2(1e12))

iterate 14 times
(1070379110497, 756872327473)


If we use `find_arr`, it will iterate at least $1070379110497 - 10^{12} = 70379110497$ times (iteration for `N + 1`) to get the answer.

## Reference:
1. https://www.mathblog.dk/project-euler-100-blue-discs-two-blue/